engines/python/setup/djl_python/rolling_batch.py

```diff
@@ -0,0 +1,226 @@
+
+from djl_python.scheduler import HuggingfaceBlock, BloomBlock, SearchConfig, SeqBatchScheduler
+
+from collections import namedtuple, defaultdict
+import torch
+
+DEFAULT_SEARCH_ALGORITHM = 'greedy'
+
+MODEL_TYPE_2_BLOCK = {'bloom': BloomBlock}
+
+
+class Request(object):
+    """
+    This class represents each request that comes to the handler.
+
+    In rolling batch, handler is called for each forward function.
+    So this class represents the states of each request until the
+    last token is generated.
+
+    """
+
+    def __init__(self, input_text: str):
+        """
+        Initialize a request
+
+        :param input_text: request's input text
+        """
+        self.input_text = input_text
+        self.next_token = None
+        self.last_token = False
+
+    def set_next_token(self, next_token: str, last_token: bool = False):
+        """
+        Sets the newly generated token.
+
+        :param next_token: next token to be set.
+        :param last_token: whether this token is the last of the sequence.
+        """
+        self.next_token = next_token
+        self.last_token = last_token
+
+    def get_next_token(self) -> str:
+        """
+        Gets the token generated for the request.
+
+        :return: next_token
+        """
+        return self.next_token
+
+    def is_last_token(self) -> bool:
+        """
+        Whether the generated token is the last one
+
+        :return: whether last token of the sequence.
+        """
+        return self.last_token
+
+
+class RollingBatch:
+    """
+    This class initializes and maintains the SequenceBatchScheduler.
+    Scheduler maintains the batch and also its search states such as past key values,
+    attention masks and position ids for each decoding strategy requests.
+
+    """
+
+    def __init__(self, model, tokenizer, config, device, properties):
+        """
+        Initializes the rolling batch scheduler.
+
+        :param model: loaded model
+        :param tokenizer: tokenizer of the model
+        :param config: configuration of the model
+        :param device: model loaded device
+        :param properties: other properties of the model, such as decoder strategy
+        """
+        self.model = model
+        self.tokenizer = tokenizer
+        self.config = config
+        self.device = device
+
+        if not self.tokenizer.pad_token:
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        lm_block_cls = MODEL_TYPE_2_BLOCK.get(self.config.model_type,
+                                              HuggingfaceBlock)
+        self.lm_block = lm_block_cls(self.model)
+        self.search_config = SearchConfig(
+            eos_token_id=self.tokenizer.eos_token,
+            pad_token_id=self.tokenizer.pad_token)
+        self.search_algorithm = properties.get('decoding_strategy',
+                                               DEFAULT_SEARCH_ALGORITHM)
+
+        self.scheduler = SeqBatchScheduler(self.lm_block,
```

Comment:
can we switch scheduler?

Reply:
Refactored. RollingBatch is an abstract class. We can implement other schedulers inheriting it.